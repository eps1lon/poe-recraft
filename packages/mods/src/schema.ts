// data/* typings
// every interface has a corresponding json file.
// json files in subdirectories have the same typing as their parent json
export type Atlas = AtlasNodeProps[];
export type CraftingBenchOptions = CraftingBenchOptionsProps[];
export type Essences = EssenceProps[];
export type Items = BaseItemTypeProps[];
export type Mods = ModProps[];

// helper
export type Tag = string;

export interface ModTypeProps {
  primary: number;
}

export interface SpawnWeightProps {
  value: number;
  tag: Tag;
}

export interface StatProps {
  id: string;
}

export interface ModProps {
  id: string;
  level: number;
  domain: number;
  name: string;
  correct_group: string;
  generation_type: number;
  stat1_min: number;
  stat1_max: number;
  stat2_min: number;
  stat2_max: number;
  stat3_min: number;
  stat3_max: number;
  stat4_min: number;
  stat4_max: number;
  stat5_min: number;
  stat5_max: number;
  mod_type: ModTypeProps;
  spawn_weights: SpawnWeightProps[];
  stats: StatProps[];
  tags: Tag[];
}

export interface WeaponTypeProps {
  critical: number;
  speed: number;
  // phys
  damage_min: number;
  damage_max: number;
  range_max: number;
}

export interface AttributeRequirementProps {
  req_str: number;
  req_dex: number;
  req_int: number;
}

export interface ArmourProps {
  armour: number;
  evasion: number;
  energy_shield: number;
}

export interface ShieldTypeProps {
  block: number;
}

export type ItemClassId =
  | 'LifeFlask'
  | 'ManaFlask'
  | 'HybridFlask'
  | 'Currency'
  | 'Amulet'
  | 'Ring'
  | 'Claw'
  | 'Dagger'
  | 'Wand'
  | 'One Hand Sword'
  | 'Thrusting One Hand Sword'
  | 'One Hand Axe'
  | 'One Hand Mace'
  | 'Bow'
  | 'Staff'
  | 'Two Hand Sword'
  | 'Two Hand Axe'
  | 'Two Hand Mace'
  | 'Active Skill Gem'
  | 'Support Skill Gem'
  | 'Quiver'
  | 'Belt'
  | 'Gloves'
  | 'Boots'
  | 'Body Armour'
  | 'Helmet'
  | 'Shield'
  | 'SmallRelic'
  | 'MediumRelic'
  | 'LargeRelic'
  | 'StackableCurrency'
  | 'QuestItem'
  | 'Sceptre'
  | 'UtilityFlask'
  | 'UtilityFlaskCritical'
  | 'Map'
  | 'Unarmed'
  | 'FishingRod'
  | 'MapFragment'
  | 'HideoutDoodad'
  | 'Microtransaction'
  | 'Jewel'
  | 'DivinationCard'
  | 'LabyrinthItem'
  | 'LabyrinthTrinket'
  | 'LabyrinthMapItem'
  | 'MiscMapItem'
  | 'Leaguestone'
  | 'PantheonSoul'
  | 'UniqueFragment'
  | 'AbyssJewel'
  | 'IncursionItem';

export interface BaseItemTypeProps {
  id: string;
  name: string;
  width: number;
  height: number;
  drop_level: number;
  inherits_from: string;
  weapon_type?: WeaponTypeProps;
  component_attribute_requirement?: AttributeRequirementProps;
  component_armour?: ArmourProps;
  shield_type?: ShieldTypeProps;
  implicit_mods: ModProps[];
  item_class: string;
  tags: Tag[];
}

export interface CraftingBenchOptionsProps {
  primary: number;
  required_level: number;
  name: string;
  crafting_bench_custom_action: number;
  sockets: number;
  socket_colours: string;
  links: number;
  item_quantity: number;
  mods_key?: number;
  costs: Array<{
    amount: number;
    base_item_type: BaseItemTypeProps;
  }>;
  mod?: ModProps;
  item_classes: string[];
  npc_master_key: number;
}

export interface WorldAreaProps {
  id: string;
  name: string;
  area_level: number;
  tags: Tag[];
  area_type_tags: Tag[];
  mods: ModProps[];
}

export interface AtlasNodeProps {
  primary: number;
  x: number;
  y: number;
  adjacent: number[];
  world_area: WorldAreaProps;
}

export interface EssenceProps {
  primary: number;
  level: number;
  base_item_type: BaseItemTypeProps;
  essence_type: {
    id: string;
    essence_type: number;
    is_corrupted_essence: boolean;
  };
  item_level_restriction: number;
  quiver_mod?: ModProps;
  amulet_mod?: ModProps;
  ring_mod?: ModProps;
  belt_mod?: ModProps;
  shield_mod?: ModProps;
  helmet_mod?: ModProps;
  body_armour_mod?: ModProps;
  boots_mod?: ModProps;
  gloves_mod?: ModProps;
  bow_mod?: ModProps;
  wand_mod?: ModProps;
  staff_mod?: ModProps;
  two_hand_sword_mod?: ModProps;
  two_hand_axe_mod?: ModProps;
  two_hand_mace_mod?: ModProps;
  claw_mod?: ModProps;
  dagger_mod?: ModProps;
  one_hand_sword_mod?: ModProps;
  one_hand_thrusting_sword_mod?: ModProps;
  one_hand_axe_mod?: ModProps;
  one_hand_mace_mod?: ModProps;
  sceptre_mod?: ModProps;
}
