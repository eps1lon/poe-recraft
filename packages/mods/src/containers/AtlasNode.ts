import Mod from '../mods/Mod';
import { AtlasNodeProps, Tag } from '../schema';

import ImmutableContainer from './ImmutableContainer';

export const SEXTANT_RANGE = 55; // http://poecraft.com/atlas has 55

export interface Builder {
  mods: Mod[];
  props: AtlasNodeProps;
}
export type HumanId = string;

export default class AtlasNode extends ImmutableContainer<Mod, Builder> {
  public static build(props: AtlasNodeProps) {
    return new AtlasNode([], props);
  }

  public static withBuilder(builder: Builder): AtlasNode {
    return new AtlasNode(builder.mods, builder.props);
  }

  public props: AtlasNodeProps;

  constructor(mods: Mod[], props: AtlasNodeProps) {
    super(mods);

    this.props = props;
  }

  public builder(): Builder {
    return {
      mods: this.mods,
      props: this.props,
    };
  }

  /**
   * returns a list of nodes with sextant distance <= depth
   * sextant distance:
   * (a, b) => 1, if a.isInSextantRange(b)
   *           +inf, otherwise
   */
  public inSextantRange(atlas: AtlasNode[], max_depth = 1): AtlasNode[] {
    let expand: AtlasNode[] = [this];
    const in_range: Set<AtlasNode> = new Set();

    let depth = max_depth;

    // with every step mark every node in range as to be expanded and
    // decrement depth
    /* eslint-disable no-loop-func */
    while (expand.length && depth >= 0) {
      expand = expand.reduce((nodes, node) => {
        in_range.add(node);

        const new_nodes = atlas.filter(
          new_node =>
            node.isInSextantRange(new_node) &&
            node !== new_node &&
            !in_range.has(new_node),
        );

        return nodes.concat(new_nodes);
      }, [] as AtlasNode[]);

      depth -= 1;
    }
    /* eslint-enable no-loop-func */

    return Array.from(in_range.values());
  }

  public isInSextantRange(other: AtlasNode): boolean {
    return this.distance(other) <= SEXTANT_RANGE;
  }

  public distance(other: AtlasNode): number {
    return Math.sqrt(
      (this.props.x - other.props.x) ** 2 + (this.props.y - other.props.y) ** 2,
    );
  }

  public pos(): string {
    return `X: ${this.props.x} Y: ${this.props.y}`;
  }

  public getAllMods(atlas: AtlasNode[]): Mod[] {
    return this.inSextantRange(atlas).reduce(
      (mods, other) => mods.concat(other.mods),
      this.mods,
    );
  }

  public getTags(): Tag[] {
    return super
      .getTags()
      .concat(this.props.world_area.tags)
      .filter(
        // unique by id
        (tag, i, tags) => tags.findIndex(other => other === tag) === i,
      );
  }

  public maxModsOfType(): number {
    return Number.POSITIVE_INFINITY;
  }

  public inDomainOf(mod_domain: number): boolean {
    return mod_domain === Mod.DOMAIN.ATLAS;
  }

  public level(): number {
    return this.props.world_area.area_level;
  }

  public affectingMods(atlas: AtlasNode[]): Mod[] {
    return atlas.reduce((mods, other) => {
      if (this.isInSextantRange(other)) {
        return mods.concat(other.mods);
      } else {
        return mods;
      }
    }, [] as Mod[]); // this.mods will be passed on atlas.reduce
  }

  public activeMods(atlas: AtlasNode[]): Mod[] {
    return this.affectingMods(atlas).filter(
      mod => mod.spawnweightFor(this) > 0,
    );
  }

  public inactiveMods(atlas: AtlasNode[]): Mod[] {
    return this.affectingMods(atlas).filter(
      mod => mod.spawnweightFor(this) <= 0,
    );
  }
}
