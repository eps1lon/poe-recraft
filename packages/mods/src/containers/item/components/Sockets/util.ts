import Item from '../../Item';
import { Socket, SocketColor } from './types';
import ValueRange from '../../../../calculator/ValueRange';

/**
 * creates a Socket according to the highest attr requirement
 * if two or more requirements are equal the color is chosen
 * with the priority Green > Red > Blue
 * @param item
 */
export function rollRequirementBiasedStable(item: Item): Socket {
  const { dex, int, str } = item.requirements.list();
  const color_by_attr = [
    { value: dex.value, color: SocketColor.green },
    { value: str.value, color: SocketColor.red },
    { value: int.value, color: SocketColor.blue },
    // descending
  ].sort((a, b) => -ValueRange.cmpMax(a.value, b.value));

  return {
    color: color_by_attr[0].color,
  };
}
