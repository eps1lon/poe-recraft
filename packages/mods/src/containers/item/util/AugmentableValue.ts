/**
 * a numeric value that was augmented in some way
 *
 * {augmented} is true if the the base value was changed to {value}
 */
export default interface AugmentableValue {
  value: number | [number, number];
  augmented: boolean;
}

export function isZero(
  augmentable: { value: AugmentableValue['value'] } | null | undefined,
): boolean {
  if (augmentable == null) {
    return true;
  }
  const { value } = augmentable;

  return (
    value === 0 || (Array.isArray(value) && value[0] === 0 && value[1] === 0)
  );
}

export function anyCmp(
  augmentable: { value: AugmentableValue['value'] },
  cmp: (n: number) => boolean,
): boolean {
  const { value } = augmentable;

  if (typeof value === 'number') {
    return cmp(value);
  } else {
    return value.some(cmp);
  }
}
