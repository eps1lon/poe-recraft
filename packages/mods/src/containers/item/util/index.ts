export { default as AugmentableValue } from './AugmentableValue';
