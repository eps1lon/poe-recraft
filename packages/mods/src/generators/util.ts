import { GeneratorDetails } from './Generator';

// filter nothing
const allowAll = () => true;

export interface WithSpawnchance {
  spawnchance: number;
}

export const calculateSpawnchance = (
  collection: Array<GeneratorDetails<any>>,
  filter: (details: GeneratorDetails<any>) => boolean = allowAll,
): Array<GeneratorDetails<any> & WithSpawnchance> => {
  const sum_spawnweight = collection.reduce((sum, details) => {
    if (details.spawnweight != null && filter(details)) {
      return sum + details.spawnweight;
    } else {
      return sum;
    }
  }, 0);

  return collection.map(details => {
    let spawnchance = 0;

    if (details.spawnweight != null && filter(details)) {
      spawnchance = details.spawnweight / sum_spawnweight;
    }

    return {
      ...details,
      spawnchance,
    };
  });
};
