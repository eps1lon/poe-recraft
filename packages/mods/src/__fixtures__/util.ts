import * as fs from 'fs';
import * as path from 'path';

import {
  AtlasNodeProps,
  BaseItemTypeProps,
  CraftingBenchOptionsProps,
  EssenceProps,
  ModProps,
} from '../schema';

import AtlasNode from '../containers/AtlasNode';
import Item from '../containers/item/Item';
import { Essence, MasterBenchOption } from '../generators';
import Mod from '../mods/Mod';
import PropsTable, { TableProps } from '../helpers/PropsTable';
import { Buildable } from '../interfaces/Buildable';

const json_cache: Partial<Tables> = {};

export const createTable = <P extends TableProps, T, A1>(
  file: keyof Tables,
  constructor: Buildable<P, T, A1>,
  arg1: A1,
) => {
  if (json_cache[file] == null) {
    const body = fs.readFileSync(
      path.join(__dirname, `../../data/${file}.json`),
    );

    const json = JSON.parse(body.toString());

    // @ts-ignore
    json_cache[file] = new PropsTable(json, constructor, arg1);
  }

  return json_cache[file];
};

interface Tables {
  atlas: PropsTable<AtlasNodeProps, AtlasNode, undefined>;
  craftingbenchoptions: PropsTable<
    CraftingBenchOptionsProps,
    MasterBenchOption,
    undefined
  >;
  essences: PropsTable<EssenceProps, Essence, ModProps>;
  items: PropsTable<BaseItemTypeProps, Item, undefined>;
  mods: PropsTable<ModProps, Mod, undefined>;
}

export const createTables = () => {
  const tables = {
    atlas: createTable('atlas', AtlasNode, undefined),
    craftingbenchoptions: createTable(
      'craftingbenchoptions',
      MasterBenchOption,
      undefined,
    ),
    items: createTable('items', Item, undefined),
    mods: createTable('mods', Mod, undefined),
  } as Tables;

  // @ts-ignore: its not undefined though?!
  tables.essences = createTable('essences', Essence, tables.mods.all());

  return tables;
};

export const byWorldAreaId = (id: string) => (node: AtlasNodeProps) =>
  node.world_area.id === id;

export const fromWorldAreaId = (
  id: string,
  atlas: PropsTable<AtlasNodeProps, AtlasNode, undefined>,
): AtlasNode => {
  return atlas.from(node => node.world_area.id === id);
};

export const byEssenceName = (name: string) => (essence: EssenceProps) =>
  essence.base_item_type.name === name;
