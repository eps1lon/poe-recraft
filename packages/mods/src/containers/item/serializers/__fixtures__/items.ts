import { createTables } from '../../../../__fixtures__/util';

const { items, mods } = createTables();

export default [
  items
    .fromName('Platinum Kris')
    .rarity.set('magic')
    .addMod(
      mods.fromId('LocalIncreasedPhysicalDamagePercentAndAccuracyRating2'),
    )
    .setProperty('mirrored', true),
  items
    .fromName('Hubris Circlet')
    .rarity.set('rare')
    .asShaperItem()
    .addMod(mods.fromId('IncreasedLife0'))
    .addMod(mods.fromId('LocalIncreasedEnergyShield2'))
    .addMod(mods.fromId('LocalIncreasedEnergyShieldPercentAndStunRecovery3'))
    .addMod(mods.fromId('FireResist1'))
    .addMod(mods.fromId('ReducedLocalAttributeRequirements1'))
    .addMod(mods.fromId('AdditionalMinesPlacedSupportedUber1_'))
    .setProperty('corrupted', true),
];
