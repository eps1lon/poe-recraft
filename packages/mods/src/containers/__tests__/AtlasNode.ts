import { createTables, fromWorldAreaId } from '../../__fixtures__/util';
import { AtlasNodeProps } from '../../schema';

import AtlasNode from '../AtlasNode';

const { atlas: atlas_nodes, mods } = createTables();

let atlas: AtlasNode[] = [];

const getNode = (id: string) => {
  const node = atlas.find(
    (other: AtlasNode) => other.props.world_area.id === id,
  );

  if (node === undefined) {
    throw new Error(`node ${id} not found`);
  }

  return node;
};
const getNodeIndex = (id: string) =>
  atlas.findIndex((node: AtlasNode) => node.props.world_area.id === id);
const ids = ({
  props: {
    world_area: { id },
  },
}: AtlasNode) => id;

beforeEach(() => {
  atlas = atlas_nodes
    .all()
    .map((node_props: AtlasNodeProps) => new AtlasNode([], node_props));
});

it('should build', () => {
  expect(fromWorldAreaId('MapWorldsDunes', atlas_nodes)).toBeInstanceOf(
    AtlasNode,
  );
});

it('should know sextant ranges', () => {
  const dunes = fromWorldAreaId('MapWorldsDunes', atlas_nodes);
  const far_map = fromWorldAreaId('MapWorldsThicket', atlas_nodes);
  const close_map = fromWorldAreaId('MapWorldsOrchard', atlas_nodes);

  expect(dunes.isInSextantRange(close_map)).toBe(true);
  expect(close_map.isInSextantRange(dunes)).toBe(true);

  expect(dunes.isInSextantRange(far_map)).toBe(false);
  expect(far_map.isInSextantRange(dunes)).toBe(false);
});

it('should collect maps in range', () => {
  const dunes = fromWorldAreaId('MapWorldsDunes', atlas_nodes);
  const dunes_unique = fromWorldAreaId('MapWorldsDunesUnique', atlas_nodes);
  const orchard = fromWorldAreaId('MapWorldsOrchard', atlas_nodes);
  const peninsula = fromWorldAreaId('MapWorldsPeninsula', atlas_nodes);
  const strand = fromWorldAreaId('MapWorldsStrand', atlas_nodes);
  const villa = fromWorldAreaId('MapWorldsVilla', atlas_nodes);

  const affected = dunes.inSextantRange([
    dunes,
    dunes_unique,
    orchard,
    peninsula,
    strand,
    villa,
  ]);

  expect(affected).toHaveLength(3);
  expect(affected).toEqual([dunes, dunes_unique, orchard]);
});

it('should be able to incrementally get maps in range', () => {
  const dunes = getNode('MapWorldsDunes');

  expect(dunes).toBeInstanceOf(AtlasNode);

  expect(dunes.inSextantRange(atlas, 0).map(ids)).toEqual(['MapWorldsDunes']);

  expect(
    dunes
      .inSextantRange(atlas, 1)
      .map(ids)
      .sort(),
  ).toMatchSnapshot();

  expect(
    dunes
      .inSextantRange(atlas, 2)
      .map(ids)
      .sort(),
  ).toMatchSnapshot();
});

it('should consider its area level', () => {
  expect(fromWorldAreaId('MapWorldsDunes', atlas_nodes).level()).toBe(74);
});

it('should know by which mods its affected', () => {
  const strand_index = getNodeIndex('MapWorldsStrand');
  const waka_index = getNodeIndex('MapWorldsStrandUnique');

  const master_mod = mods.fromId('MapAtlasContainsMaster');
  const invasion_mod = mods.fromId('MapAtlasContainsAdditionalRandomBoss');

  atlas[waka_index] = atlas[waka_index].addMod(master_mod);

  expect(atlas[waka_index].mods).toContain(master_mod);
  expect(atlas[waka_index].affectingMods(atlas)).toContain(master_mod);
  expect(atlas[waka_index].activeMods(atlas)).not.toContain(master_mod);
  expect(atlas[waka_index].inactiveMods(atlas)).toContain(master_mod);

  atlas[strand_index] = atlas[strand_index].addMod(invasion_mod);
  expect(atlas[strand_index].mods).toContain(invasion_mod);
  expect(atlas[strand_index].affectingMods(atlas)).toContain(invasion_mod);
  expect(atlas[strand_index].affectingMods(atlas)).toContain(master_mod);
  expect(atlas[waka_index].affectingMods(atlas)).toContain(invasion_mod);
  expect(atlas[waka_index].activeMods(atlas)).toContain(invasion_mod);
  expect(atlas[waka_index].inactiveMods(atlas)).not.toContain(invasion_mod);
});
