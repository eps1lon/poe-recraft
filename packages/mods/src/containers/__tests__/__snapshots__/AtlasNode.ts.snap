// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be able to incrementally get maps in range 1`] = `
Array [
  "MapWorldsDunes",
  "MapWorldsDunesUnique",
  "MapWorldsFields",
  "MapWorldsOrchard",
  "MapWorldsTropicalIsland",
]
`;

exports[`should be able to incrementally get maps in range 2`] = `
Array [
  "MapWorldsChateau",
  "MapWorldsCitySquare",
  "MapWorldsDunes",
  "MapWorldsDunesUnique",
  "MapWorldsEstuary",
  "MapWorldsFields",
  "MapWorldsInfestedValley",
  "MapWorldsLair",
  "MapWorldsOrchard",
  "MapWorldsPier",
  "MapWorldsSpiderForest",
  "MapWorldsTropicalIsland",
  "MapWorldsWastePool",
]
`;
